src/app/notes/services/notes-populator.service.ts

```diff
@@ -0,0 +1,82 @@
+import { Injectable } from '@angular/core';
+import { NotesService } from './notes.service';
+import { Note } from '../models/note.model';
+
+export enum NoteType {
+  SpecialRequirements = 'Special Requirements',
+  FacilityRequirements = 'Facility Requirements',
+  OtherNote = 'Other note'
+}
+
+@Injectable()
+export class NotesPopulatorService {
+  constructor(private readonly notesService: NotesService) {
+  }
+
+  // iterates recursively through the whole object to find where there are any placeholders for notes
+  // fetches notes and populates those placeholders
+  public async populateWithNotes(entity: any) {
+    const entityIds = this.collectEntityIds(entity, [])
+
+    this.fetchNotes(entityIds).subscribe(
```

Comment:
I don't like fetching notes from server here. NotesPopulatorService should have no awareness of http calls. I'd put the notes for given entityIds as input parameter to populateWithNotes function.